clients/src/main/java/org/apache/kafka/common/NodeVersions.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common;
+
+import org.apache.kafka.common.errors.UnsupportedVersionException;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.NavigableSet;
+import java.util.Objects;
+import java.util.TreeSet;
+
+/**
+ * A public class which represents the versions found on a node.
+ */
+public class NodeVersions {
+    /**
+     * The set of NodeVersions for this client software.
+     */
+    public static final NodeVersions CLIENT_VERSIONS;
+
+    static {
+        HashMap<Short, ApiVersionRange> apiVersionRanges = new HashMap<>();
+        for (ApiKey api : ApiKey.VALUES) {
+            apiVersionRanges.put(api.id(), api.supportedRange());
+        }
+        CLIENT_VERSIONS = new NodeVersions(apiVersionRanges);
+    }
+
+    /**
+     * A mapping from API name to version range.
+     */
+    private final Map<Short, ApiVersionRange> apiVersionRanges;
+
+    public NodeVersions(Map<Short, ApiVersionRange> apiVersionRanges) {
+        this.apiVersionRanges = Collections.unmodifiableMap(apiVersionRanges);
+    }
+
+    /**
+     * Get the ApiVersionRange for a specific API.
+     *
+     * @return                                  The ApiVersionRange.
+     * @throws UnsupportedVersionException      If no versions of the API are supported.
+     */
+    public ApiVersionRange apiVersionRange(ApiKey api) {
```

Comment:
I added a method override for `apiVersionRange(short)`.  I still think it's nice to have the ApiKey version as well.  It's always nice to work with types when you can rather than primitives... it's just kind of annoying in one or two cases because of the whole "APIs that weren't defined when the software was released" issue.